src/main/scala/org/broadinstitute/dsde/workbench/sam/service/TosService.scala

```diff
@@ -50,32 +51,56 @@ class TosService(val directoryDao: DirectoryDAO, val tosConfig: TermsOfServiceCo
+    latestUserTos <- directoryDao.getUserTos(samUser.id, samRequestContext)
+    previousUserTos <- directoryDao.getUserTosVersion(samUser.id, tosConfig.previousVersion, samRequestContext)
+    userHasAcceptedLatestVersion = userHasAcceptedLatestTosVersion(latestUserTos)
+    permitsSystemUsage = tosAcceptancePermitsSystemUsage(samUser, latestUserTos, previousUserTos)
+  } yield TermsOfServiceComplianceStatus(samUser.id, userHasAcceptedLatestVersion, permitsSystemUsage)
 
   /** If grace period enabled, don't check ToS, return true If ToS disabled, return true Otherwise return true if user has accepted ToS, or is a service account
     */
-  private def tosAcceptancePermitsSystemUsage(user: SamUser, userTos: Option[SamUserTos]): Boolean = {
-    val userIsServiceAccount = StandardSamUserDirectives.SAdomain.matches(user.email.value) // Service Account users do not need to accept ToS
-    val userIsPermitted = userTos.exists { tos =>
+  private def tosAcceptancePermitsSystemUsage(user: SamUser, userTos: Option[SamUserTos], previousUserTos: Option[SamUserTos]): Boolean = {
+    if (!tosConfig.isTosEnabled) {
+      return true
+    }
+    // Service Account users do not need to accept ToS
+    val userIsServiceAccount = StandardSamUserDirectives.SAdomain.matches(user.email.value)
+    if (userIsServiceAccount) {
+      return true
+    }
+    if (userHasRejectedLatestTosVersion(userTos)) {
+      return false
+    }
+    userTos.exists { tos =>
       val userHasAcceptedLatestVersion = userHasAcceptedLatestTosVersion(Option(tos))
       val userCanUseSystemUnderGracePeriod = tosConfig.isGracePeriodEnabled && tos.action == TosTable.ACCEPT
-      val tosDisabled = !tosConfig.isTosEnabled
 
-      userHasAcceptedLatestVersion || userCanUseSystemUnderGracePeriod || tosDisabled
+      val userHasAcceptedPreviousVersion = userHasAcceptedPreviousTosVersion(previousUserTos)
+      val now = Instant.now()
+      val userInsideOfRollingAcceptanceWindow = tosConfig.rollingAcceptanceWindowExpiration match {
+        case Some(expiration) =>
+          expiration.isAfter(now) && userHasAcceptedPreviousVersion
+        case None => false
+      }
+
+      userHasAcceptedLatestVersion || userInsideOfRollingAcceptanceWindow || userCanUseSystemUnderGracePeriod
 
     }
-    userIsPermitted || userIsServiceAccount
   }
 
   private def userHasAcceptedLatestTosVersion(userTos: Option[SamUserTos]): Boolean =
     userTos.exists { tos =>
       tos.version.contains(tosConfig.version) && tos.action == TosTable.ACCEPT
     }
+
+  private def userHasRejectedLatestTosVersion(userTos: Option[SamUserTos]): Boolean =
+    userTos.exists { tos =>
+      tos.version.contains(tosConfig.version) && tos.action == TosTable.REJECT
+    }
+
+  private def userHasAcceptedPreviousTosVersion(previousUserTos: Option[SamUserTos]): Boolean =
+    previousUserTos.exists(tos => tos.action == TosTable.ACCEPT)
```

Comment:
I think the case where there is no previous ToS needs to be included here, and if there isn't a previoius ToS, this function returns `true`

Reply:
Should probably add a test for this as well.

Reply:
This actually does return false in that case and there is a test 😄 https://github.com/broadinstitute/sam/pull/1207/files#diff-3a7b95de48385b66cf0a3b94e7e625e53bcce9efc68ce85a67cfe479f2989e1dR153

Reply:
(exists returns false if the option is none)

Reply:
Doh. You're right.